_includes/markdown/Typescript.md

```diff
@@ -0,0 +1,180 @@
+
+Note: we recommend setting the `noImplicitAny` setting to `true` in `tsconfig.json` which forces explicitly setting a type as any. This makes it easier to identify code that is relying on any and would also make catching the escape hatch in code reviews which is a great opportunity to get feedback and potentially get the types right.
+
+#### 3. The `// @ts-expect-error` comment
+
+Lastly, if nothing else worked to unblock the code, TypeScript allows disregarding errors on a given line by just telling the compiler that an error is expected on that line. This is usually the least effective option as it would only solve the problem for a single line. However, if nothing else works, this is the last and most drastic escape hatch.
+
+While we'd like none of these escape hatches to be used, it would be unrealistic to expect all engineers to have types 100% correct on every project all the time. We believe that these recommendations provide a path to address them organically during the life-cycle of the project.
+
+## TypeScript and CI
+
+We recommend that all projects run the typescript compiler to check type safety at the CI level. Every Pull Request/Merge Request should run `tsc --noEmit` on CI to ensure there are no compilation issues before merging the branch.
+
+This is a great strategy to avoid builds failing due to type errors. With Next.js for instance, you might not realize you have a TS compilation issue until a build is executed, therefore to avoid merging code that produces failing builds we recommend always ensuring that the code has no TypeScript compilation issues before merging.
+
+## Recommended `tsconfig.json` settings
+
+This is not a full list but we highlight some recommended settings here.
+
+1. `noFallthroughCasesInSwitch: true`- [Definition](https://www.typescriptlang.org/tsconfig/#noFallthroughCasesInSwitch) - ensures you are not skipping anything in a switch statement, and always at least provide a default option.
+2. `noImplicitAny: true` - [Definition](https://www.typescriptlang.org/tsconfig/#noImplicitAny) - ensures you define `any` explicitly, rather than implicitly assume it. This makes it a lot easier to find all code relying on any for future refactoring.
+3. `strictNullChecks: true` - [Definition](https://www.typescriptlang.org/tsconfig/#strictNullChecks) - forces handling both `undefined` and `null`
+4. `noUnusedParameters: true` - [Definition](https://www.typescriptlang.org/tsconfig/#noUnusedParameters) - does not allow unused parameters to remain in code
+
+## TypeScript and React
+
+### Use `type` for React Props
+
+While `interface` and `type` are mostly interchangeable, prefer `type` over `interfaces` unless you expect the props to be the base for other components's props.
+
+```ts
+type BlocksProps = {
+	html: string;
+};
+
+export const Blocks = ({ html }: BlocksProps) => {
```

Comment:
Wouldn't `string`, `undefined` be a valid `JSX.Element`? so does it really matter? I'm not necessarily against it just haven't found it to be very useful. But I'm open to it if it makes sense.

@kasparszarinovs any thoughts here?

Reply:
They would not. Things like `string`, `null`, `undefined` would not be valid JSX.Element. https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/react/index.d.ts#L327

My personal preference is to always add return types over having TS infer them but I'm not that bothered by it either if we don't

Reply:
I would agree with @tobeycodes  that return types are important. We actually tend to enforce them on the Node side (not yet on FE) - this helps ensure that your code works as expected.

I may be a bit out of the loop as to why `React.FC` is not favoured - so please do let me know what I'm missing. Personally, however, I don't see n issue with using those.

tl;dr; - yes, do use return types if you can :)

Reply:
I've used to use React.FC a lot but a few things happened:

1. Until React 17 (I think) it would always include `children` even if you didn't need it.
2. Since React 18 it does not include `children` anymore
3. Most of the times we don't need the other things from React.FC so it's a littler easier for beginners not have to think about what React.FC is
4. We likely have projects still using React < 18, which could potentially run into a confusion if deciding to use TS before upgrading React (and its types packages).

Since React 18 and recent versions of TS though there's [nothing wrong with it](https://www.totaltypescript.com/you-can-stop-hating-react-fc) (altought some still prefer just typing props), I just feel like it could be confusing given its history and projects on older versions of react.

However between typing return types and just using React.FC I can see React.FC being better, perhaps we just need to add a note about it. In any case, I feel like we would need a new section since I'm not even mentioning either one in the doc today :)

Regarding return types I know it's a hot debate - I personally haven't found issues with most functions/components. Of course in some cases you should definately add a return type. So I tend to only type when needed.

Reply:
I'll draft up a new section around `Typing React Components` and `Return Types`. If you have any suggestions/wording feel free to drop a comment here.

Reply:
I personally think this is a great example of where React core team did something, then realized it was a bad decision and then made a breaking change to fix it. I personally think we should make decisions based on the latest release of a framework. The beauty of TS here is that users are going to know immediately whether their FC type is correct whether they use children or not because of the feedback they get in their IDE or at build time. 

If projects have the freedom they should use TS first before upgrading React. That is going to help highlight any breaking changes automatically when they upgrade and FC would be one example of this where types need to change (which can be completely automated with codemods)

I also think using FC is great for beginners because they are able to do a Find by Reference and see what else a function component can do.

My 2 cents on return types is the more engineers do it the faster they will learn TS, but also we shouldn't force it

But at the end of the day, I'm not really that opinionated on all of this. The path of least resistance for most engineers is best

Reply:
On some projects we've used this convention:

```
const Header: React.FC<ComponentProps> = () => {}
```

It was a nice convention, Im not sure it holds any value of any other approach but I found it more readable in general.